core/trino-spi/src/main/java/io/trino/spi/Page.java

```diff
@@ -274,22 +274,23 @@ public void compact()
      */
     public Page getLoadedPage()
     {
-        Block[] loadedBlocks = null;
         for (int i = 0; i < blocks.length; i++) {
             Block loaded = blocks[i].getLoadedBlock();
             if (loaded != blocks[i]) {
-                if (loadedBlocks == null) {
-                    loadedBlocks = blocks.clone();
+                // Transition to new block creation mode after the first newly loaded block is encountered
+                Block[] loadedBlocks = new Block[blocks.length];
```

Comment:
Why not just `clone()` here? Does the optimization to only copy the subset up to `i` actually matter?

Reply:
If loading does occur, then a `clone()` will have this new `Block[] loadedBlocks` pointing into the old blocks (potentially tenured GC regions) which requires card marking- whereas by storing the loaded blocks into the array as they are loaded has a better chance of the array and the new blocks belonging to the same region (which doesn't require the card marking). I think this effect is very minor compared to the benefit of removing the branching path from the point after which blocks are loaded.

Reply:
It that's the case and we can't prove that's actually a problem, I'd opt for the simpler code (i.e., just clone and remove the System.arrayCopy block)

Reply:
I think at this point now that the loop is re-structured to "transition to load blocks" the removal of `clone()` is more importantly traversing `loadedBlocks` between 1-2 times (ie: `blocks[0]` triggers loading, then `blocks[1..n]` are traversed directly best case, worst case `block[n - 1]` triggers loading and we traverse again) instead of always traversing the array twice via `clone()`, which is just hiding the full copying traversal underneath the built-in method.

Reply:
Right, however my question is whether that matters in practice. This is an operation that happens at most once per page and it's in the order of number of columns, vs everything else the engine already has to do per row.

Also, allocating the array also has to walk the array to initialize all cells to `null`, so I'm not sure it really is saving anything.

Reply:
An observation though- Pages with a large number of columns also tend to have a very small number of rows per page because of the page size target that PageBuilder and page splitting use- so the “once per page” cost is disproportionately paid when it matters more.

Incidentally, the initialization of the array with nulls is something that C2 can elide if it can prove that all values are assigned and that uninitialized values are never read- but admittedly this optimization is typically characterized as “fragile”.

Anyway, I have no data to prove that my current implementation achieves either array initialization or outperforms clone()  (but it might!) so I’m fine switching this back to clone().

Reply:
> An observation though- Pages with a large number of columns also tend to have a very small number of rows per page because of the page size target that PageBuilder and page splitting use- so the “once per page” cost is disproportionately paid when it matters more.

If we have degenerated pages with few rows and large amount of columns, IMO we should think about merging them, etc. Sure, column enumeration shouldn't be the worst performing code, but it's not as optimized as row processing.